apps/zotonic_core/src/support/z_module_manager.erl

```diff
@@ -634,21 +660,138 @@ module_exists(M) ->
+    }.
+mod_info(Module) ->
+    App = module_to_app(Module),
+    Mod = module_to_mod(App),
+    LibDir = case code:lib_dir(App) of
+        {error, _} -> undefined;
+        Dir -> unicode:characters_to_binary(Dir)
+    end,
+    #{
+        app => App,
+        prio => prio(Mod),
+        version => mod_version(App),
+        schema => mod_schema(App),
+        title => mod_title(App),
+        description => mod_description(App),
+        app_dir => LibDir
+    }.
+
+-spec mod_version(Module) -> Version when
+    Module :: atom() | binary() | string(),
+    Version :: binary().
+mod_version(Module) ->
+    App = module_to_app(Module),
+    case application:get_key(App, vsn) of
+        {ok, ""} ->
+            app_git_version(App);
+        {ok, "git"} ->
+            app_git_version(App);
+        {ok, Vsn} ->
+            unicode:characters_to_binary(Vsn, utf8);
+        undefined ->
+            app_git_version(App)
+    end.
+
+app_git_version(App) ->
+    case maybe_git_version(code:lib_dir(App)) of
+        undefined ->
+            maybe_git_version(z_path:site_source_dir(App));
+        Vsn ->
+            Vsn
+    end.
+
+maybe_git_version({error, _}) ->
+    undefined;
+maybe_git_version(LibDir) ->
+    GitDir = filename:join(LibDir, ".git"),
+    case filelib:is_dir(GitDir) of
+        true ->
+            git_version(LibDir);
+        false ->
+            undefined
+    end.
+
+git_version(LibDir) ->
+    Cmd = "git rev-parse --short HEAD",
+    case exec:run(Cmd, [sync, stdout, {cd, LibDir}]) of
+        {ok, Res} ->
+            {stdout, Hash} = proplists:lookup(stdout, Res),
+            iolist_to_binary([
+                "git-", z_string:trim(unicode:characters_to_binary(Hash))
+            ]);
+        {error, Reason} ->
+            ?LOG_WARNING(#{
+                in => zotonic_core,
+                text => <<"Git rev-parse for module version failed">>,
+                cmd => unicode:characters_to_binary(Cmd),
+                lib_dir => LibDir,
+                result => error,
+                reason => Reason
+            }),
+            undefined
+    end.
 
 %% @doc Get the title of a module.
-title(M) ->
+-spec mod_title(Module) -> Title when
+    Module :: atom() | binary() | string(),
+    Title :: binary() | undefined.
+mod_title(Module) ->
+    Mod = module_to_mod(Module),
     try
-        proplists:get_value(mod_title, M:module_info(attributes))
+        Title = proplists:get_value(mod_title, Mod:module_info(attributes), <<>>),
+        unicode:characters_to_binary(Title)
     catch
         _M:_E -> undefined
     end.
 
+%% @doc Get the description of a module.
+-spec mod_description(Module) -> Desc when
+    Module :: atom() | binary() | string(),
+    Desc :: binary() | undefined.
+mod_description(Module) ->
+    Mod = module_to_mod(Module),
+    try
+        Desc = proplists:get_value(mod_description, Mod:module_info(attributes), <<>>),
+        unicode:characters_to_binary(Desc)
```

Comment:
Should author etc also use `unicode:characters_to_binary`?

Reply:
Good point.  With something extra, as the author is sometimes an atom in Erlang modules...